tuner/bayes_opt_gp/dffml_tuner_bayes_opt_gp/bayes_opt_gp.py

```diff
@@ -0,0 +1,166 @@
+from typing import Union, Dict, Any, List
+import itertools
+import logging
+
+from dffml.base import (
+    config,
+    field,
+)
+from dffml.noasync import train, score
+from dffml.tuner import Tuner, TunerContext
+from dffml.util.entrypoint import entrypoint
+from dffml.record import Record
+from dffml.source.source import BaseSource
+from dffml.accuracy import AccuracyScorer, AccuracyContext
+from dffml.model import ModelContext
+from dffml.feature import Feature
+import nest_asyncio
+from bayes_opt import BayesianOptimization
+
+
+class InvalidParametersException(Exception):
+    pass
+
+
+@config
+class BayesOptGPConfig:
+    parameters: dict = field(
+        "Parameters to be optimized", default_factory=lambda: dict()
+    )
+    objective: str = field(
+        "How to optimize the given scorer. Values are min/max", default="max"
+    )
+    init_points: int = field(
+        "How many steps of random exploration you want to perform.", default=5
+    )
+    n_iter: int = field(
+        "How many steps of bayesian optimization you want to perform.",
+        default=10,
+    )
+
+
+class BayesOptGPContext(TunerContext):
+    """
+    Bayesian Optimization GP Tuner
+    """
+
+    async def optimize(
+        self,
+        model: ModelContext,
+        feature: Feature,
+        accuracy_scorer: Union[AccuracyScorer, AccuracyContext],
+        train_data: Union[BaseSource, Record, Dict[str, Any]],
+        test_data: Union[BaseSource, Record, Dict[str, Any]],
+    ):
+        """
+        Method to optimize hyperparameters by Bayesian optimization using Gaussian Processes
+        as the surrogate model.
+        Uses a grid of hyperparameters in the form of a dictionary present in config,
+        Trains each permutation of the grid of parameters and compares accuracy.
+        Sets model to the best parameters and returns highest accuracy.
+
+        Note that for this tuner, each hyperparameter field to be tuned must have exactly 2 values 
+        specified, representing the minimum and maximum values in the search space for that 
+        hyperparameter. Additionally, they must be either float/integer values. Otherwise,
+        an error is raised.
+
+        Parameters
+        ----------
+        model : ModelContext
+            The Model which needs to be used.
+
+        feature : Feature
+            The Target feature in the data.
+
+        accuracy_scorer: AccuracyContext
+            The accuracy scorer that needs to be used.
+
+        train_data: SourcesContext
+            The train_data to train models on with the hyperparameters provided.
+
+        sources : SourcesContext
+            The test_data to score against and optimize hyperparameters.
+
+        Returns
+        -------
+        float
+            The highest score value
+        """
+
+        nest_asyncio.apply()
```

Comment:
```suggestion
```

Reply:
This is probably something for a different PR, but it could potentially go well at the start of the `noasync` functions.

Reply:
Since the train/score functions in the asynchronous tune function need to be synchronous, nest_asyncio is necessary. Shall I move it to the noasync file then?